### PR TITLE
Queues - Danielle Birbal & Tamiko Terada - Scrabble

### PR DESCRIPTION
**Pull request by birbalds**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|  We were able to distribute the responsibilities well by weaving in pair-programming concepts with our new git collaboration commands. We found ourselves finishing early for each wave which gave us time to expand, refine and explore. 	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  This is a self method because we don’t need a new instance of ‘Scoring’ each time a score needs to be created using the method. Scoring is something that is consistent for all players in the game so it does not need to be created each time.  |
|   Describe an example of a test you wrote for a _nominal case_. |   We created ArgumentError tests to make sure that a word could not be scored if it has non-alpha characters (numbers/symbols) or any spaces.	|
|   Describe an example of a test you wrote for an _edge case_. |   We tested the case for when a player has exactly 100 points (= minimum extreme to win the game) so that we could see if the .play method would return ‘false’ instead of allowing the player to continue playing the game.	|
|   What was your final test coverage percentage?	|   100% coverage for all of our required .rb files (no tests created for board.rb). 	|
|   Is there a specific piece of code you'd like feedback on?	|   We would like feedback on our tilebag_spec.rb test for 'removes the correct letters from the set'	|

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            1 => %w(A E I O U L N R S T),
+            2 => %w(D G),
+            3 => %w(B C M P),
+            4 => %w(F H V W Y),
+            5 => %w(K),
+            8 => %w(J X),
+            10 => %w(Q Z)
+        }.freeze
+
+        def score(word)
```
Comment (droberts-sea) [inline lib/scoring.rb:14]: This should be a self method! Define it like this:

```ruby
def self.score(word)
end
```

And use it like this

```ruby
Scoring.score("test)
```

That way you don't even need an instance of `Scoring`, which doesn't really make sense anyhow.

File: lib/scoring.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                @letter = letter.upcase
+                SCORE_CHART.each do |number, array|
+                    @word_score += number if array.include?(@letter)
+                end
+            end
+            @word_score
+        end
+
+        def highest_score_from(array_of_words)
+            @all_scores = array_of_words.map { |word| score(word) }
```
Comment (droberts-sea) [inline lib/scoring.rb:35]: I'm curious why you chose to use instance variables (variables starting with an `@`) here. It seems to me that none of these values need to be saved between calls to this method, so local variables would be more appropriate.

Basically, you should always use a local variable (no `@`) unless you've got a good reason to do something else.

File: lib/scoring.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+            @scored_words = Hash[array_of_words.zip(@all_scores)]
+
+            @high_words = @scored_words.find_all { |_word, score| @scored_words.values.max == score }
+
+            @highest_word = @high_words.min_by do |word|
+                if word[0].length == 7
+                    @high_words.index(word)
```
Comment (droberts-sea) [inline lib/scoring.rb:43]: Why do you use the index here? That would indicate that if a word has 7 letters, you treat it as though it had as many letters as it's position in the array, which doesn't make much sense.

For example,

```ruby
highest_score_from(['zzzzzz', 'qzzzzz', 'zqzzzz', 'zzqzzz', 'zzzqzz', 'zzzzqz', 'zzzzzq', 'qqzzzz', 'zqqzzz', 'aaaaaah'])
```

returns `"zzzzzz"` (6 letters, 60 points) when it should be `"aaaaaah"` (7 letters, 60 points).

Instead, since we know 7-letter words are supposed to win, you could assign it an impossibly low length like 0.

File: lib/scoring.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        }.freeze
+
+        def score(word)
+            # starting word_score is based on word.length
+            if word.length < 7 && !word.empty?
+                @word_score = 0
+            elsif word.length == 7
+                @word_score = 50
+            else
+                raise ArgumentError, 'invalid word length'
```
Comment (droberts-sea) [inline lib/scoring.rb:21]: Good defensive programming.

File: lib/scoring.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
+module Scrabble
+    class Scoring
+        attr_accessor :word_score
+        SCORE_CHART = {
```
Comment (droberts-sea) [inline lib/scoring.rb:4]: This hash of arrays definitely works, but looking up a letter's score ends up being complicated, involving an `each` loop.

Instead, what if you switched it around and had the keys in this array be letters, and the values be the corresponding score. So it would look like

```ruby
SCORE_CHART = {
  'a': 1,
  'b': 3,
  # ...
}
```

Then looking up a letter would be as simple as `SCORE_CHART[letter]`.

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
+require_relative './scoring'
+require_relative './tilebag'
+
+module Scrabble
+    class Player < Scoring
```
Comment (droberts-sea) [inline lib/player.rb:5]: Why does `Player` inherit from `Scoring`? I don't think that the "is-a" relationship holds here - I don't think of a `Player` as being a type of `Scoring` (in fact that doesn't really make sense).

I think that composition is probably the relationship you want here - `Player` _calls_ `Scoring.score` as an external method when it needs something scored.

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        end
+        it 'returns sum of scores of played words' do
+            %w(z zzz zz zqaa).each { |word| @new_player.play(word) }
+            @new_player.total_score.must_equal 82
+        end
+    end
+
+    describe '#won?' do
+        it 'returns #won as a Boolean value' do
+            [true, false].must_include @new_player.won?
```
Comment (droberts-sea) [inline specs/player_spec.rb:92]: This is a nitpick, but I don't think this test is necessary. Since all values in Ruby are either truthy or falsey, explicitly checking whether something is `true` or `false` is a little overkill.

File: specs/player_spec.rb [hunk 8]
```diff
@@ -0,0 +1,155 @@
+require 'simplecov'
+SimpleCov.start
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require 'minitest/skip_dsl'
+require_relative '../lib/player'
+
+describe Scrabble::Player do
+    before do
```
Comment (droberts-sea) [inline specs/player_spec.rb:10]: Nice use of `before` to DRY up your code!

File: lib/tilebag.rb [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+        def draw_tiles(number)
+            raise ArgumentError, 'not enough tiles' if @in_bag.length < number
+            @tiles_drawn = @in_bag.pop(number)
+            @tiles_drawn
+        end
+
+        def tiles_remaining
+            if @in_bag.empty?
+                puts 'There are no more tiles!'
```
Comment (droberts-sea) [inline lib/tilebag.rb:31]: Seems like you ought to just return 0 in this case, rather than `puts`ing something to the terminal.

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   yes	|
|   Both Teammates contributed to the codebase	|   yes	|
|   Regular Commits with meaningful commit messages 	|   yes	|
|  **Scoring class**	|   	|
|  `score` method	|   yes	|
|   Uses appropriate data structure to store the letter score	|   see inline comment	|
|   Appropriately handles edge / failure cases	|   yes - good work on this	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   7-letter yes, 1-letter no	|
|  `highest_scoring_word` method	|   yes	|
|   Appropriately handles edge cases	|   mostly - see inline comment	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   some - would like to see more coverage of ties, and don't see any coverage of other edge cases (empty array, etc)	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   yes	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   yes	|
|  Returns `true` or `false` from `won?` method |  yes |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |   yes  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   yes	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   yes	|
|  `tiles_remaining` method returns a count	|   yes	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   yes	|

Good work overall! This is a big project, which means I'm always going to have a lot of nitpicks. However, in general I'm seeing lots of good habits - clean, readable code, solid test coverage, good git hygiene. Keep it up!

The one thing I'm worried about is there still seems to be some confusion about instance vs local variables, and instance vs class methods. It might be worth reviewing that textbook material and/or finding some online resources about it.
